### DISCUSSION TITLE
ADD CUSTOM CLAIMS DYNAMICALLY

### DISCUSSION THREAD

**Discussion by ismailatitallah** [keycloak/keycloak#17279] [Q&A]
Hello,
Currently I am using a technical account to generate token from keycloack.
Then I need to add additional information to the generated token from the connected user. (for specefic needs, we can't declare the user in the keycloak users, that's why we pass by a technical account)
Those piece of information are dynamically constructed (based on the connected user who ordered the token generation)

I found this documentation : 

`If you want to add a custom claim to a Keycloak token when a user is authenticated, you can use the Keycloak REST API to modify the user's token. Specifically, you can use the /auth/realms/{realm}/protocol/openid-connect/token endpoint to request a new token with custom claims included.
Here's an example of how to use the Keycloak REST API to add a custom claim to a user's token:
First, you'll need to authenticate the user and obtain an access token. This can be done using the Keycloak authentication API, for example, by redirecting the user to the Keycloak login page or using the Resource Owner Password Credentials grant. The access token will be included in the response from Keycloak.
Once you have the access token, you can use it to make a request to the /auth/realms/{realm}/protocol/openid-connect/token endpoint to request a new token with custom claims. The request should include the following parameters:
grant_type: This should be set to refresh_token.
client_id: The ID of the client application that the token is being requested for.
refresh_token: The refresh token obtained in step 1.
scope: The scope of the new token, which can include any additional scopes required by the client application.
custom_claims: A JSON object containing the custom claims to be added to the token.`

But It doesn't work. 
My Question is there a solution to add dynamic information to the token via api?

Thank you very much

**Comment by vplemyannik** [comment]
You can implement your own custom SPI 
```java
public class MyCustomOIDCProtocolMapper extends AbstractOIDCProtocolMapper implements OIDCAccessTokenMapper, OIDCIDTokenMapper, UserInfoTokenMapper {
    public static final String PROVIDER_ID = "oidc-custom-protocol-mapper";
    private static final List<ProviderConfigProperty> configProperties = new ArrayList<>();


    @Override
    public List<ProviderConfigProperty> getConfigProperties() {
        return configProperties;
    }

    @Override
    public String getDisplayCategory() {
        return TOKEN_MAPPER_CATEGORY;
    }

    @Override
    public String getDisplayType() {
        return "My custom claim mapper";
    }

    @Override
    public String getId() {
        return PROVIDER_ID;
    }

    @Override
    public String getHelpText() {
        return "Add additional claims dynamically";
    }


    // this code will be invoked for each token request whether it's access_token or refresh_token request 
    @Override
    public AccessToken transformAccessToken(
            AccessToken token,
            ProtocolMapperModel mappingModel,
            KeycloakSession keycloakSession,
            UserSessionModel userSession,
            ClientSessionContext clientSessionCtx) {


        // here your logic for getting claims 
       // it might be http call or db query or whatever you need 
      
        Map<String, Object> claims = token.getOtherClaims();
        claims.put("MyClaim", "MyClaimValue");

        setClaim(token, mappingModel, userSession, keycloakSession, clientSessionCtx);
        return token;

```

**Reply by Laurenzi** [reply]
I have an use case with Client Credentials Grant where I want to pass a custom parameter (with a dynamic value) that could be read by the custom claims mapper to decide which claims to map to the access token.
It would be hugely beneficial to be able to access the parameters / Client Credentials Grant request body in the mapper also.

Does anyone have any idea how to go on about implementing such mapper?

**Reply by justin-tay** [reply]
Why do you need a custom parameter? Can't you just use the scope parameter?

**Reply by Laurenzi** [reply]
@justin-tay 
Well it'd be of course better if I can just use scope parameter. 
Do you happen to have idea about how that scope parameter can be mapped to claims, if I need some custom processing upon the scope value?

**Reply by justin-tay** [reply]
Typically you configure a client scope, the mappers for mapping claims for that scope, then configure the client to have that optional/default scope.

**Reply by Laurenzi** [reply]
@justin-tay to illustrate my scenario, consider the following:
1. I want to obtain access token with some dynamic value based on the client grant request (ie. the request must contain some information that is then mapped to the access token)
2. Let's say the Client Credentials Grant request body looks like this: grant_type=client_credentials&custom_param_to_be_mapped=some-dynamic-value-the-token-obtainer-can-define
3. Now exactly how would I make it so that my token mapper would map value of "customer_param_to_be_mapped" to the access token? I don't mind if I have to use some other parameter name, that is not the issue here. What I mean is the token requestor is able to define a dynamic value that would be part of the access token to be granted.

Hope this sorted out what I was after. I am not aware that there is an existing way to do that.

**Reply by justin-tay** [reply]
I don't really get how dynamic this value is. The claim is typically for some form of authorization so it can't be that any client can put any value there without some form of validation. So then for each value you would create a scope. And for each scope you would have a hardcoded claim mapper with the value. Then for each eligible client you would configure the scopes it can set.

**Reply by Laurenzi** [reply]
Well, with Authorization Code Grant, it's entirely possible to write a custom implementation of `Authenticator` to introduce some dynamic lookup mechanism for the claims that are put to access token for example, based on what was presented in the authorization grant request.

The thing here is different in a sense that I am talking about Client Credentials Grant, which to my understanding, does not provide a mean to override used `Authenticator`. Or if it does, then I am not simply aware of such choice.

**Reply by justin-tay** [reply]
I'm still not really sure about your use case. It could be that you need something like [RFC 9396 OAuth 2.0 Rich Authorization Requests](https://datatracker.ietf.org/doc/html/rfc9396) with the `authorization_details` parameter but that doesn't look like it's been implemented in Keycloak. One option then is to enable the experimental `dynamic-scopes` feature where you can specify a dynamic scope with static and variable component.

![dynamicscope](https://github.com/keycloak/keycloak/assets/49700559/4a755c4e-b340-4e69-9e95-2ed4c6fddc91)

Then your scope parameter can be `mycustomscope:some-dynamic-value-the-token-obtainer-can-define`.

In your custom `AbstractOIDCProtocolMapper`

```java
@Override
protected void setClaim(final IDToken token,
                        final ProtocolMapperModel mappingModel,
                        final UserSessionModel userSession,
                        final KeycloakSession keycloakSession,
                        final ClientSessionContext clientSessionCtx) {
    clientSessionCtx.getScopeString(); // returns "mycustomscope:some-dynamic-value-the-token-obtainer-can-define"
}
```

**Reply by Laurenzi** [reply]
Thanks @justin-tay that was exactly what I was looking for!

**Reply by kisamoto** [reply]
Sorry for the delayed repost but does this work for you?

I have dynamic scopes set up in Keycloak 26.3.2 however in my custom `setClaim` `OIDCProtocolMapper` I am not seeing the scopes sent by the client. I know my mapper is being triggered (so the scope is being requested correctly) however `getScopeString()` is only showing `openid profile email`.
